# Limpeza e Tratamento dos Dados

Nesse notebook iremos fazer a limpeza e o tratamento do dataset que iremos usar nesse projeto. Após o processo de limpeza e tratamento, iremos salvar um novo dataset no formato parquet.

## Importando as Bibliotecas

Primeiro, vamos importar as bibliotecas que usaremos nessa etapa.

In [58]:
import numpy as np
import pandas as pd

## Carregando o Dataset

Agora, vamos carregar o dataset e ver algumas informações gerais.

In [59]:
df = pd.read_csv("../data/raw/zap_dataset.csv")

In [60]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10118 entries, 0 to 10117
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id                  10118 non-null  int64  
 1   titulo              10118 non-null  str    
 2   tipo_imovel         10118 non-null  str    
 3   url                 10118 non-null  str    
 4   preco               10116 non-null  float64
 5   moeda               10116 non-null  str    
 6   condominio          5822 non-null   float64
 7   area_m2             10118 non-null  int64  
 8   quartos             10117 non-null  float64
 9   banheiros           10117 non-null  float64
 10  vagas               9245 non-null   float64
 11  bairro              4065 non-null   str    
 12  rua                 9427 non-null   str    
 13  cidade              10118 non-null  str    
 14  estado              10118 non-null  str    
 15  pais                10118 non-null  str    
 16  aceita_pets    

## Tratando e Limpando as Features

Agora, vamos analisar feature por feature, tratar e limpar seus valores.

In [61]:
def get_info(column, values=True):
    num_na = column.isna().sum()
    num_total = len(column)

    print(f"dtype: {column.dtype}")
    print(f"NA: {num_na}/{num_total} ({100*num_na/num_total:.2f}%)")

    counts = column.value_counts(dropna=False)

    if values:
        df_values = pd.DataFrame({
            "contagem": counts,
            "porcentagem": 100*counts/num_total
        })

        df_values["porcentagem"] = (df_values["porcentagem"]
            .map(lambda x: f"{x:.2f} %")
        )
        
        print(f"valores:\n{df_values}")

Mas antes, vamos checar se há registros duplicados.

In [62]:
df[df.duplicated()].info()

<class 'pandas.DataFrame'>
Index: 352 entries, 8160 to 8752
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id                  352 non-null    int64  
 1   titulo              352 non-null    str    
 2   tipo_imovel         352 non-null    str    
 3   url                 352 non-null    str    
 4   preco               352 non-null    float64
 5   moeda               352 non-null    str    
 6   condominio          244 non-null    float64
 7   area_m2             352 non-null    int64  
 8   quartos             352 non-null    float64
 9   banheiros           352 non-null    float64
 10  vagas               318 non-null    float64
 11  bairro              161 non-null    str    
 12  rua                 346 non-null    str    
 13  cidade              352 non-null    str    
 14  estado              352 non-null    str    
 15  pais                352 non-null    str    
 16  aceita_pets         

Como podemos ver, existem registros duplicados. Portanto vamos removê-los.

In [63]:
df = (df
    .drop_duplicates()
    .reset_index(drop=True)
)

In [64]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 9766 entries, 0 to 9765
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id                  9766 non-null   int64  
 1   titulo              9766 non-null   str    
 2   tipo_imovel         9766 non-null   str    
 3   url                 9766 non-null   str    
 4   preco               9764 non-null   float64
 5   moeda               9764 non-null   str    
 6   condominio          5578 non-null   float64
 7   area_m2             9766 non-null   int64  
 8   quartos             9765 non-null   float64
 9   banheiros           9765 non-null   float64
 10  vagas               8927 non-null   float64
 11  bairro              3904 non-null   str    
 12  rua                 9081 non-null   str    
 13  cidade              9766 non-null   str    
 14  estado              9766 non-null   str    
 15  pais                9766 non-null   str    
 16  aceita_pets      

### Feature `id`

In [65]:
get_info(df["id"])

dtype: int64
NA: 0/9766 (0.00%)
valores:
            contagem porcentagem
id                              
2902564623         1      0.01 %
2835720338         1      0.01 %
2857048455         1      0.01 %
2811993532         1      0.01 %
2901920002         1      0.01 %
...              ...         ...
2899359088         1      0.01 %
2900444692         1      0.01 %
2852856061         1      0.01 %
2892982315         1      0.01 %
2870426627         1      0.01 %

[9766 rows x 2 columns]


Como podemos ver, a feature apenas representa os id's dos imóveis. Como essa informação não estará presente durante a inferência, então podemos remover ela.

In [66]:
df = df.drop("id", axis="columns")

### Feature `titulo`

In [67]:
get_info(df["titulo"])

dtype: str
NA: 0/9766 (0.00%)
valores:
                                                    contagem porcentagem
titulo                                                                  
Flat à venda com 1 quarto, 1 suíte, 1 banheiro ...        76      0.78 %
Apartamento à venda com 2 quartos, 1 suíte, 2 b...        51      0.52 %
Flat à venda com 1 quarto, 1 banheiro e 1 vaga ...        36      0.37 %
Apartamento à venda com 1 quarto, 1 suíte, 1 ba...        33      0.34 %
Flat à venda com 1 quarto, 1 suíte, 1 banheiro ...        28      0.29 %
...                                                      ...         ...
Apartamento à venda com 2 quartos, 2 suítes, 72...         1      0.01 %
Apartamento para comprar com 44 m², 1 quarto, 1...         1      0.01 %
Apartamento à venda com 3 quartos, 3 suítes, de...         1      0.01 %
Apartamento à venda com 3 quartos, 3 suítes, 4 ...         1      0.01 %
Apartamento à venda com 3 quartos, 1 suíte, 77 ...         1      0.01 %

[6820 rows 

Essa feature contem os títulos dos anúncios dos imóveis e obviamente não estará disponível na inferência, assim como a feature `id`. Entretanto, como ela fornece muitas informações sobre o imóvel em si, podemos usar ela para validar as informações das outras features.

### Feature `tipo_imovel`

In [68]:
get_info(df["tipo_imovel"])

dtype: str
NA: 0/9766 (0.00%)
valores:
             contagem porcentagem
tipo_imovel                      
House            4570     46.80 %
Apartment        3289     33.68 %
Place            1907     19.53 %


Essa feature indica o tipo de imóvel (casa, apartamento, etc). Nela vamos apenas renomear as classes.

In [69]:
df["tipo_imovel"] = df["tipo_imovel"].map({
    "House": "casa",
    "Apartment": "apartamento",
    "Place": "outro",
})

Além disso, podemos usar a feature `titulo` ou `descricao_completa` para gerar outras classes para `tipo_imovel`.

### Feature `url`

In [70]:
get_info(df["url"])

dtype: str
NA: 0/9766 (0.00%)
valores:
                                                    contagem porcentagem
url                                                                     
https://www.zapimoveis.com.br/imovel/venda-apar...         1      0.01 %
https://www.zapimoveis.com.br/lancamentos/smart...         1      0.01 %
https://www.zapimoveis.com.br/imovel/venda-apar...         1      0.01 %
https://www.zapimoveis.com.br/imovel/venda-apar...         1      0.01 %
https://www.zapimoveis.com.br/imovel/venda-apar...         1      0.01 %
...                                                      ...         ...
https://www.zapimoveis.com.br/imovel/venda-apar...         1      0.01 %
https://www.zapimoveis.com.br/imovel/venda-apar...         1      0.01 %
https://www.zapimoveis.com.br/imovel/venda-apar...         1      0.01 %
https://www.zapimoveis.com.br/imovel/venda-apar...         1      0.01 %
https://www.zapimoveis.com.br/imovel/venda-apar...         1      0.01 %

[9766 rows 

Essa feature contem as URLs dos anúncios dos imóveis. Como não teremos ela na inferência, então vamos removê-la.

In [71]:
df = df.drop("url", axis="columns")

### Feature `preco` (Target)

In [72]:
get_info(df["preco"])

dtype: float64
NA: 2/9766 (0.02%)
valores:
           contagem porcentagem
preco                          
450000.0        209      2.14 %
550000.0        197      2.02 %
650000.0        165      1.69 %
750000.0        126      1.29 %
350000.0        124      1.27 %
...             ...         ...
882604.0          1      0.01 %
1579999.0         1      0.01 %
1668280.0         1      0.01 %
1216000.0         1      0.01 %
1591679.0         1      0.01 %

[1480 rows x 2 columns]


Como podemos ver, os valores de preço já estão no tipo `float64` e aparentam estar na unidade monetária correta (Reais R$). Agora, podemos notar que existem dois valores ausentes de preço, portanto devemos removê-los.

In [73]:
df = df.dropna(subset="preco")

### Feature `moeda`

In [74]:
get_info(df["moeda"])

dtype: str
NA: 0/9764 (0.00%)
valores:
       contagem porcentagem
moeda                      
BRL        9764    100.00 %


Como podemos notar, essa feature indica a moeda usada no anúncio e como todos os anúncios foram feitos no Brasil, obviamente a moeda seria Reais. Como essa feature tem variância zero, ela é inútil e portanto será removida.

In [75]:
df = df.drop("moeda", axis="columns")

### Feature `condominio`

In [76]:
get_info(df["condominio"])

dtype: float64
NA: 4186/9764 (42.87%)
valores:
            contagem porcentagem
condominio                      
NaN             4186     42.87 %
1.0              534      5.47 %
500.0            259      2.65 %
800.0            258      2.64 %
1000.0           256      2.62 %
...              ...         ...
774.0              1      0.01 %
3004.0             1      0.01 %
514.0              1      0.01 %
12.0               1      0.01 %
1355.0             1      0.01 %

[586 rows x 2 columns]


Assim como `preco`, a feature `condominio` já está no tipo `float64` e também possui valores faltantes. Entretanto, diferente de `preco` não precisamos remover os registros com valores faltantes. Portanto, não faremos nada por enquanto.

### Feature `area_m2`

In [77]:
get_info(df["area_m2"])

dtype: int64
NA: 0/9764 (0.00%)
valores:
         contagem porcentagem
area_m2                      
150           151      1.55 %
120           126      1.29 %
200           118      1.21 %
100           117      1.20 %
160           115      1.18 %
...           ...         ...
533             1      0.01 %
979             1      0.01 %
630             1      0.01 %
524             1      0.01 %
585             1      0.01 %

[539 rows x 2 columns]


Como podemos notar, essa feature também já está bem polida, não faremos nada agora.

### Feature `quartos`

In [78]:
get_info(df["quartos"])

dtype: float64
NA: 1/9764 (0.01%)
valores:
         contagem porcentagem
quartos                      
3.0          4729     48.43 %
2.0          1891     19.37 %
4.0          1677     17.18 %
1.0          1146     11.74 %
5.0           288      2.95 %
6.0            27      0.28 %
7.0             4      0.04 %
NaN             1      0.01 %
8.0             1      0.01 %


Assim como a feature `area_m2`, não precisaremos fazer nada por enquanto.

### Feature `banheiros`

In [79]:
get_info(df["banheiros"])

dtype: float64
NA: 1/9764 (0.01%)
valores:
           contagem porcentagem
banheiros                      
2.0            2850     29.19 %
1.0            1847     18.92 %
4.0            1737     17.79 %
3.0            1344     13.76 %
5.0            1145     11.73 %
6.0             607      6.22 %
7.0             153      1.57 %
8.0              60      0.61 %
9.0              15      0.15 %
NaN               1      0.01 %
23.0              1      0.01 %
22.0              1      0.01 %
10.0              1      0.01 %
12.0              1      0.01 %
17.0              1      0.01 %


Novamente, não faremos nada por enquanto.

### Feature `vagas`

In [80]:
get_info(df["vagas"])

dtype: float64
NA: 838/9764 (8.58%)
valores:
       contagem porcentagem
vagas                      
2.0        3314     33.94 %
1.0        2348     24.05 %
4.0        1675     17.15 %
3.0        1102     11.29 %
NaN         838      8.58 %
6.0         226      2.31 %
5.0         189      1.94 %
8.0          24      0.25 %
7.0          20      0.20 %
9.0           9      0.09 %
12.0          6      0.06 %
20.0          4      0.04 %
10.0          3      0.03 %
26.0          2      0.02 %
11.0          1      0.01 %
14.0          1      0.01 %
16.0          1      0.01 %
18.0          1      0.01 %


Também não faremos nada nessa feature.

### Feature `bairro`

In [81]:
get_info(df["bairro"])

dtype: str
NA: 5862/9764 (60.04%)
valores:
                                      contagem porcentagem
bairro                                                    
NaN                                       5862     60.04 %
Setor Bueno                                683      7.00 %
Setor Marista                              336      3.44 %
Setor Oeste                                192      1.97 %
Parque Amazônia                            186      1.90 %
...                                        ...         ...
Residencial Bertim Belchior 1ª etapa         1      0.01 %
Residencial Antônio Barbosa                  1      0.01 %
Loteamento Grande Retiro                     1      0.01 %
Vila Anchieta                                1      0.01 %
Vila Matilde                                 1      0.01 %

[329 rows x 2 columns]


Essa feature representa o bairro onde o imóvel está localizado e será útil na etapa de enriquecimento do dataset. Por hora, não faremos nada.

### Feature `rua`

In [82]:
get_info(df["rua"])

dtype: str
NA: 685/9764 (7.02%)
valores:
                            contagem porcentagem
rua                                             
NaN                              685      7.02 %
Rua T 37                         215      2.20 %
Rua T 36                         159      1.63 %
Rua T 55                         145      1.49 %
Rua T 30                         132      1.35 %
...                              ...         ...
Rua Amaral Peixoto                 1      0.01 %
Rua RSL 13                         1      0.01 %
Rua Monsenhor Primo Vieira         1      0.01 %
Praça Central                      1      0.01 %
Rua Célia C. S. de Brito           1      0.01 %

[2099 rows x 2 columns]


Assim como a feature `bairro`, não faremos nada.

### Feature `cidade`

In [83]:
get_info(df["cidade"])

dtype: str
NA: 0/9764 (0.00%)
valores:
         contagem porcentagem
cidade                       
Goiânia      9764    100.00 %


Como todos os anúncios foram extraídos de Goiânia, então isso justifica a variância zero da feature. Como ela não terá relevância preditiva, iremos removê-la.

In [84]:
df = df.drop("cidade", axis="columns")

### Feature `Estado`

In [85]:
get_info(df["estado"])

dtype: str
NA: 0/9764 (0.00%)
valores:
        contagem porcentagem
estado                      
GO          9764    100.00 %


Como ocorreu com `cidade`, essa feature não tem variância e consequentemente será removida.

In [86]:
df = df.drop("estado", axis="columns")

### Feature `pais`

In [87]:
get_info(df["pais"])

dtype: str
NA: 0/9764 (0.00%)
valores:
      contagem porcentagem
pais                      
BR        9764    100.00 %


Novamente, iremos remover essa feature pela falta de variabilidade.

In [88]:
df = df.drop("pais", axis="columns")

### Feature `aceita_pets`

In [89]:
get_info(df["aceita_pets"])

dtype: bool
NA: 0/9764 (0.00%)
valores:
             contagem porcentagem
aceita_pets                      
False            6386     65.40 %
True             3378     34.60 %


Essa feature já está bem tratada, não faremos nada.

### Feature `comodidades`

In [90]:
get_info(df["comodidades"])

dtype: str
NA: 1697/9764 (17.38%)
valores:
                                                    contagem porcentagem
comodidades                                                             
NaN                                                     1697     17.38 %
Garage                                                   158      1.62 %
Pets Allowed                                              33      0.34 %
Barbecue Grill                                            28      0.29 %
Gated Community                                           21      0.22 %
...                                                      ...         ...
Green Space, Safety Circuit, Cinema, Heating, N...         1      0.01 %
Gym, Service Area, Exterior View, Garden, Gourm...         1      0.01 %
Barbecue Grill, Internet Access, Gated Communit...         1      0.01 %
Gated Community, Gourmet Kitchen, Intercom, Nea...         1      0.01 %
Safety Circuit, Caretaker, Sports Court, Servic...         1      0.01 %

[7379 r

Essa feature parece ter bastante informação relevante, mas não faremos nada com ela por enquanto.

### Feature `fotos_urls`

In [91]:
get_info(df["fotos_urls"])

dtype: str
NA: 0/9764 (0.00%)
valores:
                                                    contagem porcentagem
fotos_urls                                                              
https://resizedimgs.zapimoveis.com.br/img/vr-li...         9      0.09 %
https://resizedimgs.zapimoveis.com.br/img/vr-li...         9      0.09 %
https://resizedimgs.zapimoveis.com.br/img/vr-li...         7      0.07 %
https://resizedimgs.zapimoveis.com.br/img/vr-li...         6      0.06 %
https://resizedimgs.zapimoveis.com.br/img/vr-li...         6      0.06 %
...                                                      ...         ...
https://resizedimgs.zapimoveis.com.br/img/vr-li...         1      0.01 %
https://resizedimgs.zapimoveis.com.br/img/vr-li...         1      0.01 %
https://resizedimgs.zapimoveis.com.br/img/vr-li...         1      0.01 %
https://resizedimgs.zapimoveis.com.br/img/vr-li...         1      0.01 %
https://resizedimgs.zapimoveis.com.br/img/vr-li...         1      0.01 %

[9074 rows 

Assim como a feature `url`, essa feature contem um conjunto de URLs de fotos dos imóveis. Como processamento de imagens e visão computacional está fora do escopo desse projeto, vamos remover essa feature.

In [92]:
df = df.drop("fotos_urls", axis="columns")

### Feature `descricao_completa`

In [93]:
get_info(df["descricao_completa"])

dtype: str
NA: 6/9764 (0.06%)
valores:
                                                    contagem porcentagem
descricao_completa                                                      
CASA NOVA JARDIM PLANALTO \n\nCASA COM 3 SUÍTES...        14      0.14 %
Na melhor localização do Setor Marista na Aveni...        12      0.12 %
Flat à venda no World Trade Center - Setor Mari...        12      0.12 %
Casa Térrea no Portal do Sol II \nSe você procu...         9      0.09 %
Sobrado no Portal do Sol I\n\nBelo sobrado no P...         9      0.09 %
...                                                      ...         ...
Este apartamento está localizado no bairro Parq...         1      0.01 %
Apartamento de 1 quarto com Suíte na divisa do ...         1      0.01 %
<b>ARTE SQUARE - APARTAMENTOS DE 3 QUARTOS, 03 ...         1      0.01 %
Apartamento pronto para morar no Cristal Tower,...         1      0.01 %
Apartamento de 3 quartos sendo 1 suíte , locali...         1      0.01 %

[8650 rows 

Essa feature contem as descrições dos anúncios dos imóveis no zap imóveis e como ela não estará presente na inferência, então podemos removê-la.

In [94]:
df = df.drop("descricao_completa", axis="columns")

## Validando o Dataset

Nessa etapa iremos validar algumas informações do dataset com a feature `titulo`.

In [95]:
quartos = df["titulo"].str.extract(r"(?P<quarto>\d+) quarto[s]")["quarto"].astype("float64")
banheiros = df["titulo"].str.extract(r"(?P<banheiro>\d+) banheiro[s]")["banheiro"].astype("float64")
vagas = df["titulo"].str.extract(r"(?P<vaga>\d+) vaga[s]")["vaga"].astype("float64")
area_m2 = df["titulo"].str.extract(r"(?P<area>\d+) m²")["area"].astype("float64")

df_validate = pd.DataFrame({
    "quartos": quartos,
    "banheiros": banheiros,
    "vagas": vagas,
    "area_m2": area_m2,
})

df_validate

,quartos,banheiros,vagas,area_m2
0,2.0,2.0,NaN,60.0
1,NaN,NaN,NaN,68.0
2,3.0,5.0,2.0,125.0
3,3.0,4.0,4.0,244.0
4,2.0,2.0,NaN,57.0
...,...,...,...,...
9761,3.0,2.0,NaN,79.0
9762,NaN,NaN,NaN,44.0
9763,3.0,5.0,3.0,174.0
9764,3.0,4.0,2.0,123.0


In [96]:
df_values = df[["quartos", "banheiros", "vagas", "area_m2"]].values.reshape(-1)
df_validate_values = df_validate.values.reshape(-1)
mask = pd.Series(df_validate_values).notna()

correct = (df_values[mask] == df_validate_values[mask]).sum()
total = mask.sum()

print(f"Porcentagem de dados coerentes: {100*correct / total:.2f}%")

Porcentagem de dados coerentes: 97.98%


Como podemos ver, das informações presentes no título do anúncio cerca de $97.98\%$ estão coerentes com os dados extraídos, o que representa uma alta taxa de integridade.

## Convertendo os `dtypes`

Agora, antes de finalizar, iremos converter os `dtypes` do nosso dataset.

In [97]:
df = df.convert_dtypes()

In [98]:
df.info()

<class 'pandas.DataFrame'>
Index: 9764 entries, 0 to 9765
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   titulo       9764 non-null   string 
 1   tipo_imovel  9764 non-null   string 
 2   preco        9764 non-null   Int64  
 3   condominio   5578 non-null   Int64  
 4   area_m2      9764 non-null   Int64  
 5   quartos      9763 non-null   Int64  
 6   banheiros    9763 non-null   Int64  
 7   vagas        8926 non-null   Int64  
 8   bairro       3902 non-null   string 
 9   rua          9079 non-null   string 
 10  aceita_pets  9764 non-null   boolean
 11  comodidades  8067 non-null   string 
dtypes: Int64(6), boolean(1), string(5)
memory usage: 991.7 KB


## Salvando o Dataset

Para finalizar, vamos salvar nosso dataset limpo em formato parquet.

In [99]:
df.to_parquet("../data/processed/01_cleaned.parquet")